File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int findPivot(vector<int>&nums){
        int l=0,r=nums.size()-1;

        while(l<r){
            int mid=l+(r-l)/2;
            // Chote elements right side h
            if(nums[mid]>nums[r]) l=mid+1;
            else r=mid;
        }
        
        return r;
    }
    int binarySearch(vector<int>&nums,int l,int r,int target){
        while(l<=r){
            int mid=l+(r-l)/2;
            if(nums[mid]==target) return mid;
            else if(nums[mid]>target) r=mid-1;
            else l=mid+1;
        }
        return -1;
    }
    int search(vector<int>& nums, int target) {
        int minIdx=findPivot(nums);

        if(nums[minIdx]==target) return minIdx;

        int idx=-1;
        idx=binarySearch(nums,0,minIdx-1,target); 
        if(idx!=-1) return idx;
        
        return binarySearch(nums,minIdx,nums.size()-1,target);
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution(object):
    def binarySearch(self, nums, target, l, r):
        while l<=r:
            m = l + (r-l) // 2
            mid = nums[m]
            if(target == mid):
                return m
            elif target > mid:
                l = m + 1
            else:
                r = m - 1
        return -1


    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        pivot = 0
        i = 0
        for i in range(n - 1):
            if nums[i] > nums[i+1]:
                pivot = i + 1
                break
        # If array not rotated, pivot will remain 0       
        
        if target >= nums[0]:
            # search before pivot
            return self.binarySearch(nums, target, 0, pivot-1 if pivot > 0 else n - 1)
        else:
            # search after pivot
            return self.binarySearch(nums, target, pivot, n-1)
            
